### ISSUE TITLE
KDA 公式推导详解

### ISSUE THREAD

**Issue by lingebeng** [lingebeng/op_dev#3]
# 1. KDA 交互矩阵 $A$ 的物理意义：从递推到矩阵方程

## Delta Rule 递推公式

对于 chunk 内的每个位置 $t$：

$$S'_t = S_{t-1} \cdot e^{g_t} \quad \text{(decay)}$$

$$e_t = v_t - \mathbf{k}_t^\top S'_t \quad \text{(prediction error)}$$

$$S_t = S'_t + \beta_t \cdot \mathbf{k}_t \otimes e_t \quad \text{(delta update)}$$

### 位置 0

$$S'_0 = S_{init} \cdot e^{g_0} = 0$$

$$e_0 = v_0 - \mathbf{k}_0^\top S'_0$$

$$S_0 = S'_0 + \beta_0 \cdot \mathbf{k}_0 \cdot e_0$$

### 位置 1

$$S'_1 = S_0 \cdot e^{g_1} = \beta_0 \cdot \mathbf{k}_0 \cdot e_0 \cdot e^{g_1}$$

$$e_1 = v_1 - \mathbf{k}_1^\top S'_1 = v_1 - \beta_0 \cdot (\mathbf{k}_1 \cdot \mathbf{k}_0) \cdot e^{g_1} \cdot e_0$$

### 位置 2

$$S'_2 = S_1 \cdot e^{g_2} = \left[\beta_0 \cdot \mathbf{k}_0 \cdot e_0 \cdot e^{g_1} + \beta_1 \cdot \mathbf{k}_1 \cdot e_1\right] \cdot e^{g_2}$$

$$e_2 = v_2 - \mathbf{k}_2^\top S'_2 = v_2 - \beta_0 \cdot (\mathbf{k}_2 \cdot \mathbf{k}_0) \cdot e^{g_1+g_2} \cdot e_0 - \beta_1 \cdot (\mathbf{k}_2 \cdot \mathbf{k}_1) \cdot e^{g_2} \cdot e_1$$

> **关键**：$e_2$ 中对位置 1 的依赖用的是 $e_1$（不是 $v_1$），而 $e_1$ 本身又依赖 $e_0$。这就是级联依赖。

## 定义交互系数 $L_{ci}$

位置 $c$ 对位置 $i$ 的交互强度（$c > i$）：

$$L_{ci} = \beta_i \cdot (\mathbf{k}_c \cdot \mathbf{k}_i) \cdot e^{G_c - G_i} \quad (d_{ci} = e^{G_c - G_i} \text{ 为从 } i \text{ 到 } c \text{ 的累积衰减})$$

代入展开式：

$$e_0 = v_0$$

$$e_1 = v_1 - L_{10} \cdot e_0$$

$$e_2 = v_2 - L_{20} \cdot e_0 - L_{21} \cdot e_1$$

## 写成矩阵方程 $(I + L) \mathbf{e} = \mathbf{v}$

$$\begin{pmatrix} 1 & 0 & 0 \\\\ L_{10} & 1 & 0 \\\\ L_{20} & L_{21} & 1 \end{pmatrix} \begin{pmatrix} e_0 \\\\ e_1 \\\\ e_2 \end{pmatrix} = \begin{pmatrix} v_0 \\\\ v_1 \\\\ v_2 \end{pmatrix}$$

其中：

$$L_{10} = \beta_0 \cdot (\mathbf{k}_1 \cdot \mathbf{k}_0) \cdot e^{G_1 - G_0}$$

$$L_{20} = \beta_0 \cdot (\mathbf{k}_2 \cdot \mathbf{k}_0) \cdot e^{G_2 - G_0}$$

$$L_{21} = \beta_1 \cdot (\mathbf{k}_2 \cdot \mathbf{k}_1) \cdot e^{G_2 - G_1}$$

## $A$ 矩阵就是 $L$

代码中构建的交互矩阵（乘 $\beta$ 并 mask 后）就是 $L$ 的下三角部分。求解过程：

$$\mathbf{e} = (I + L)^{-1} \mathbf{v}$$

通过前向替代计算 $(I + L)^{-1}$，将串行递推转化为一次矩阵乘法。

## $L_{ci}$ 的物理意义

$$L_{ci} = \underbrace{\beta_i}_{\text{位置 } i \text{ 的学习率}} \cdot \underbrace{(\mathbf{k}_c \cdot \mathbf{k}_i)}_{\text{key 相似度}} \cdot \underbrace{e^{G_c - G_i}}_{\text{从 } i \text{ 到 } c \text{ 的衰减}}$$

- **$\beta_i$ 大**：位置 $i$ 对 state 的写入强，影响后续位置更多
- **$\mathbf{k}_c \cdot \mathbf{k}_i$ 大**：两个位置的 key 越相似，位置 $c$ 越能"感知到"位置 $i$ 的写入
- **$e^{G_c - G_i}$ 小**（$c$ 远离 $i$）：距离越远衰减越大，影响越弱

## 为什么 $e_2$ 依赖 $e_1$ 而不是 $v_1$？

从递推展开：

$$S_1 = S'_1 + \beta_1 \cdot \mathbf{k}_1 \cdot e_1$$

位置 1 写入 state 的内容是 $\beta_1 \cdot \mathbf{k}_1 \cdot e_1$（修正后的预测误差），**不是** $v_1$（原始值）。

所以位置 2 通过 state 看到的是 $e_1$，这就形成了**级联依赖**：

- $e_0 = v_0$（独立）
- $e_1$ 依赖 $e_0$
- $e_2$ 依赖 $e_0$ 和 $e_1$

这正是需要求解 $(I + L)^{-1}$ 的原因——如果所有 $e$ 都独立，直接并行计算就行了，不需要矩阵求逆。

# 2. KDA 交互矩阵 $A$ 的构建过程

## 对应代码

```python
for i in range(BT):
    k_i = k[..., i, :]         # [B, H, NT, K]
    g_i = g[..., i:i + 1, :]   # [B, H, NT, 1, K]
    A = A.at[..., i].set(
        jnp.einsum('...cd,...d->...c', k * jnp.exp(g - g_i), k_i)
    )
```

> **注意**：此处 `g` 是 cumsum 后的累积 gate $G$，不是原始 gate。

## 输入（以 $BT=3, K=3$ 为例）

$A$ 的 shape 为 $[BT, BT]$，初始全 0：

$$A = \begin{pmatrix} 0 & 0 & 0 \\\\ 0 & 0 & 0 \\\\ 0 & 0 & 0 \end{pmatrix}$$

$\mathbf{k}$ 的 shape 为 $[BT, K]$：

$$\mathbf{k} = \begin{pmatrix} k_{00} & k_{01} & k_{02} \\\\ k_{10} & k_{11} & k_{12} \\\\ k_{20} & k_{21} & k_{22} \end{pmatrix}$$

$G$ 为 cumsum 后的累积 gate，shape 为 $[BT, K]$：

$$G = \begin{pmatrix} G_{00} & G_{01} & G_{02} \\\\ G_{10} & G_{11} & G_{12} \\\\ G_{20} & G_{21} & G_{22} \end{pmatrix}$$

### $i=0$：填第 0 列

取出第 0 行：

$$\mathbf{k}_i = [k_{00}, k_{01}, k_{02}], \quad \mathbf{g}_i = [G_{00}, G_{01}, G_{02}] \quad (\text{shape } [1, K]\text{，广播})$$

计算 $\mathbf{k} \odot \exp(G - \mathbf{g}_i)$（逐元素）：

$$\begin{pmatrix} k_{00} \cdot e^{G_{00}-G_{00}} & k_{01} \cdot e^{G_{01}-G_{01}} & k_{02} \cdot e^{G_{02}-G_{02}} \\\\ k_{10} \cdot e^{G_{10}-G_{00}} & k_{11} \cdot e^{G_{11}-G_{01}} & k_{12} \cdot e^{G_{12}-G_{02}} \\\\ k_{20} \cdot e^{G_{20}-G_{00}} & k_{21} \cdot e^{G_{21}-G_{01}} & k_{22} \cdot e^{G_{22}-G_{02}} \end{pmatrix}$$

与 $\mathbf{k}_i$ 做点积（沿 $K$ 维求和），得到第 0 列：

$$A_{00} = \sum_{d} k_{0d} \cdot e^{G_{0d}-G_{0d}} \cdot k_{0d} = \sum_{d} k_{0d}^2$$

$$A_{10} = \sum_{d} k_{1d} \cdot e^{G_{1d}-G_{0d}} \cdot k_{0d}$$

$$A_{20} = \sum_{d} k_{2d} \cdot e^{G_{2d}-G_{0d}} \cdot k_{0d}$$

### $i=1$：填第 1 列

$$\mathbf{k}_i = [k_{10}, k_{11}, k_{12}], \quad \mathbf{g}_i = [G_{10}, G_{11}, G_{12}]$$

$$A_{01} = \sum_{d} k_{0d} \cdot e^{G_{0d}-G_{1d}} \cdot k_{1d}$$

$$A_{11} = \sum_{d} k_{1d} \cdot e^{G_{1d}-G_{1d}} \cdot k_{1d} = \sum_{d} k_{1d}^2$$

$$A_{21} = \sum_{d} k_{2d} \cdot e^{G_{2d}-G_{1d}} \cdot k_{1d}$$

### $i=2$：填第 2 列

$$\mathbf{k}_i = [k_{20}, k_{21}, k_{22}], \quad \mathbf{g}_i = [G_{20}, G_{21}, G_{22}]$$

$$A_{02} = \sum_{d} k_{0d} \cdot e^{G_{0d}-G_{2d}} \cdot k_{2d}$$

$$A_{12} = \sum_{d} k_{1d} \cdot e^{G_{1d}-G_{2d}} \cdot k_{2d}$$

$$A_{22} = \sum_{d} k_{2d} \cdot e^{G_{2d}-G_{2d}} \cdot k_{2d} = \sum_{d} k_{2d}^2$$

## 最终 $A$ 矩阵

通用公式：

$$A_{ci} = \sum_{d=0}^{K-1} k_{cd} \cdot e^{G_{cd} - G_{id}} \cdot k_{id} \quad (d \text{ 遍历 key 的每个维度，共 } K \text{ 个})$$

$$A = \begin{pmatrix} \sum_d k_{0d}^2 & \sum_d k_{0d} e^{G_{0d}-G_{1d}} k_{1d} & \sum_d k_{0d} e^{G_{0d}-G_{2d}} k_{2d} \\\\ \sum_d k_{1d} e^{G_{1d}-G_{0d}} k_{0d} & \sum_d k_{1d}^2 & \sum_d k_{1d} e^{G_{1d}-G_{2d}} k_{2d} \\\\ \sum_d k_{2d} e^{G_{2d}-G_{0d}} k_{0d} & \sum_d k_{2d} e^{G_{2d}-G_{1d}} k_{1d} & \sum_d k_{2d}^2 \end{pmatrix}$$

## 矩阵性质

| 区域 | 条件 | $e^{G_{cd} - G_{id}}$ | 含义 |
|------|------|----------------------|------|
| 对角线 | $c = i$ | $= 1$ | key 自身模长平方 $\sum_d k_{cd}^2$ |
| 下三角 | $c > i$ | $< 1$ | 正向衰减，是有意义的交互 |
| 上三角 | $c < i$ | $> 1$ | 反向（无物理意义），Step 2 中被 mask 清零 |

- 与 $`K=1`$ 的区别：$`K=1`$ 时 $A_{ci}$ 是标量乘法；$`K>1`$ 时是**向量点积**，gate 在每个 $`K`$ 维度上可以独立衰减。

## 后续步骤

1. $A \leftarrow A \cdot \text{diag}(\boldsymbol{\beta})_{\text{row}}$ — 行方向乘 $\beta_c$
2. $A \leftarrow \text{where}(\text{mask}, 0, -A)$ — 零化上三角，取反下三角
3. 前向替代 — 解 $(I+L)^{-1}$
4. $A \leftarrow (A + I) \cdot \text{diag}(\boldsymbol{\beta})_{\text{col}}$ — 加单位阵，列方向乘 $\beta_i$

# 3. KDA 前向替代求逆：从 $-L$ 到 $(I+L)^{-1}$

## 目标

已知下三角交互矩阵 $L$，需要求解：

$$\mathbf{e} = (I + L)^{-1} \mathbf{v}$$

代码通过**前向替代**计算 $(I+L)^{-1}$，将串行递推转化为一次矩阵乘法。

## 数学推导

定义 $M = (I+L)^{-1} - I$（逆矩阵的非对角部分）。

由 $(I+L)(I+M) = I$ 展开：

$$I + M + L + LM = I$$

$$M + L + LM = 0$$

$$M = -L - LM$$

### 为什么不是死循环？

$L$ 是**严格下三角**矩阵（对角线为 0），所以 $M$ 也是严格下三角。第 $i$ 行的 $M$ 只依赖前 $i-1$ 行（已经求出），可以**逐行从上往下解**。

### 逐行展开（$BT=3$）

$$L = \begin{pmatrix} 0 & 0 & 0 \\\\ L_{10} & 0 & 0 \\\\ L_{20} & L_{21} & 0 \end{pmatrix}$$

**行 0**：$M_{0,:} = 0$（无依赖）

**行 1**：

$$M_{10} = -L_{10} - (LM)_{10} = -L_{10} - L_{10} \cdot M_{00} = -L_{10} - 0 = -L_{10}$$

**行 2**：

$$(LM)_{20} = L_{20} \cdot M_{00} + L_{21} \cdot M_{10} = 0 + L_{21} \cdot (-L_{10}) = -L_{21} L_{10}$$

$$M_{20} = -L_{20} - (LM)_{20} = -L_{20} + L_{21} L_{10}$$

$$M_{21} = -L_{21} - (LM)_{21} = -L_{21} - (L_{20} \cdot M_{01} + L_{21} \cdot M_{11}) = -L_{21}$$

### 结果

$$M = \begin{pmatrix} 0 & 0 & 0 \\\\ -L_{10} & 0 & 0 \\\\ -L_{20} + L_{21}L_{10} & -L_{21} & 0 \end{pmatrix}$$

$$\boxed{(I+L)^{-1} = I + M = \begin{pmatrix} 1 & 0 & 0 \\\\ -L_{10} & 1 & 0 \\\\ -L_{20} + L_{21}L_{10} & -L_{21} & 1 \end{pmatrix}}$$

### $M_{20}$ 的含义

$$M_{20} = \underbrace{-L_{20}}_{\text{直接路径 } 2 \leftarrow 0} + \underbrace{L_{21} L_{10}}_{\text{间接路径 } 2 \leftarrow 1 \leftarrow 0}$$

间接路径**部分抵消**了直接路径：位置 1 已经修正过位置 0 的影响，位置 2 不需要重复修正。

### 注意：代码中的 $A$ 不是数学上的 $L$

代码在 Step 1 中将 $\beta_c$（行方向）乘入了交互矩阵：

$$A_{\text{code}}[c,i] = \beta_c \cdot (\mathbf{k}_c \cdot \mathbf{k}_i) \cdot e^{G_c - G_i}$$

而数学上的 $L$ 是 $\beta_i$（列方向）：

$$L_{ci} = \beta_i \cdot (\mathbf{k}_c \cdot \mathbf{k}_i) \cdot e^{G_c - G_i}$$

代码在最后一步通过列方向乘 $\beta_i$ 来补偿这个差异。最终结果等价。

### Step 1：初始化 $A = -L'$

```python
A = jnp.where(mask, 0, -A)
```

将上三角（含对角线）置 0，下三角取反。此时 $A$ 包含 $-L'$（直接路径信息）：

$$A = \begin{pmatrix} 0 & 0 & 0 \\\\ -A_{10} & 0 & 0 \\\\ -A_{20} & -A_{21} & 0 \end{pmatrix}$$

### Step 2：前向替代循环

```python
for i in range(1, BT):
    update = jnp.einsum('...m,...mj->...j', A[..., i, :], A[..., :, :i])
    A = A.at[..., i, :i].set(A[..., i, :i] + update)
```

逐行补上间接路径的修正：

**$i=1$**：

$$\text{update}_0 = \sum_m A_{1m} \cdot A_{m0} = (-A_{10}) \cdot 0 = 0$$

$$A_{10} \mathrel{+}= 0 \quad \text{（不变）}$$

**$i=2$**：

$$\text{update}_0 = \sum_m A_{2m} \cdot A_{m0} = (-A_{20}) \cdot 0 + (-A_{21}) \cdot (-A_{10}) + 0 \cdot (-A_{20}) = A_{21} \cdot A_{10}$$

$$A_{20} \mathrel{+}= A_{21} \cdot A_{10} \quad \text{（间接路径修正）}$$

$$\text{update}_1 = \sum_m A_{2m} \cdot A_{m1} = 0$$

$$A_{21} \mathrel{+}= 0 \quad \text{（不变）}$$

**循环原理**：处理第 $i$ 行时，$A$ 的前 $i$ 行已经是 $M$ 的正确值。所以 `A[..., :, :i]` 就是 $M[:, :i]$，而 `A[..., i, :]` 还是原始的 $-L'_i$。update 计算的正好是 $-(L'M)_{i,:i}$，与 $M = -L' - L'M$ 完全对应。

### Step 3：加 $I$ 并乘 $\beta_i$

```python
A = (A + jnp.eye(BT)) * beta[..., None, :]
```

$$A_{\text{final}} = (I + M) \cdot \text{diag}(\boldsymbol{\beta})_{\text{col}}$$

- 加 $I$：从 $M = (I+L')^{-1} - I$ 变成 $(I+L')^{-1}$
- 乘 $\beta_i$（列方向）：补偿 Step 1 中 $\beta_c$（行方向）与数学 $L$（$\beta_i$ 列方向）的差异，同时使输出为 $\beta \cdot \mathbf{e}$

### 最终效果

$$\mathbf{u} = A_{\text{final}} \cdot \mathbf{v} = \boldsymbol{\beta} \odot \mathbf{e}$$

输出 $\mathbf{u}$ 是 $\beta \cdot e$（预测误差已乘学习率），后续状态更新可直接使用，无需再乘 $\beta$。

## 为什么只有下三角矩阵能这样求逆？

下三角矩阵的依赖是**单向的**（行 $i$ 只依赖行 $0, \ldots, i-1$），因此可以逐行递推求解。

对于一般矩阵，$M = -L - LM$ 中各行互相依赖，无法逐行求解。

这恰好对应 delta rule 的**因果性**：位置 $c$ 只能看到 $c$ 之前的位置，交互矩阵天然是下三角。
